django/db/models/fields/tuple_lookups.py

```diff
@@ -0,0 +1,314 @@
+
+        # e.g.: (a, b, c) <= (x, y, z) as SQL:
+        # WHERE a < x OR (a = x AND (b < y OR (b = y AND (c < z OR c = z))))
+        cols = self.lhs.get_cols()
+        lookups = itertools.cycle([LessThan, Exact])  # <, =, <, =, ...
+        connectors = itertools.cycle([OR, AND])  # OR, AND, OR, AND, ...
+        cols_list = [col for col in cols for _ in range(2)]
+        vals_list = [val for val in self.rhs for _ in range(2)]
+        cols_iter = iter(cols_list)  # a, a, b, b, c, c
+        vals_iter = iter(vals_list)  # x, x, y, y, z, z
+        col, val = next(cols_iter), next(vals_iter)
+        lookup, connector = next(lookups), next(connectors)
+        root = node = WhereNode([lookup(col, val)], connector=connector)
+
+        for col, val in zip(cols_iter, vals_iter):
+            lookup, connector = next(lookups), next(connectors)
+            child = WhereNode([lookup(col, val)], connector=connector)
+            node.children.append(child)
+            node = child
+
+        return root.as_sql(compiler, connection)
+
+    def as_postgresql(self, compiler, connection):
+        return self.tuple_comparison(compiler, connection, LessThanOrEqual)
+
+    def as_sqlite(self, compiler, connection):
+        return self.tuple_comparison(compiler, connection, LessThanOrEqual)
+
+    def as_mysql(self, compiler, connection):
+        return self.tuple_comparison(compiler, connection, LessThanOrEqual)
+
+
+class TupleIn(TupleLookupMixin, In):
+    def check_tuple_lookup(self):
+        assert isinstance(self.lhs, Cols)
+        self.check_rhs_is_tuple_or_list()
+        self.check_rhs_is_collection_of_tuples_or_lists()
+        self.check_rhs_elements_length_equals_lhs_length()
+
+    def as_sql(self, compiler, connection):
+        from django.db.models.sql.where import AND, OR, WhereNode
+
+        if not self.rhs:
+            raise EmptyResultSet
+
+        # e.g.: (a, b, c) in [(x1, y1, z1), (x2, y2, z2)] as SQL:
+        # WHERE (a = x1 AND b = y1 AND c = z1) OR (a = x2 AND b = y2 AND c = z2)
+        root = WhereNode([], connector=OR)
+        cols = self.lhs.get_cols()
+
+        for vals in self.rhs:
+            lookups = [Exact(col, val) for col, val in zip(cols, vals)]
+            root.children.append(WhereNode(lookups, connector=AND))
+
+        return root.as_sql(compiler, connection)
+
+    def tuple_in(self, compiler, connection):
+        if not self.rhs:
+            raise EmptyResultSet
+
+        # e.g.: (a, b, c) in [(x1, y1, z1), (x2, y2, z2)] as SQL:
+        # WHERE (a, b, c) IN ((x1, y1, z1), (x2, y2, z2))
+        rhs = []
+        for vals in self.rhs:
+            rhs.append(
+                Tuple(
+                    *[
+                        Value(val, output_field=col.output_field)
+                        for col, val in zip(self.lhs, vals)
+                    ]
+                )
+            )
+
+        lookup = In(Tuple(self.lhs), Tuple(*rhs))
+        return lookup.as_sql(compiler, connection)
```

Comment:
We should make and effort to handle non-direct rhs as well. Users could pass expressions that generate tuples instead such as `Subquery`. You should have a look at how the `In` lookup is implemented for not-direct values. A test could be something like

```python
Row.objects.filter(
    pk__in=Row.objects.values("id", "tenant_id")
)

Reply:
I'll take a look into this tomorrow

Reply:
The thing is I considered this before, but there's no way to reasonably support this unless the backend supports `(a, b, c) IN SELECT (x, y, z)` or `(a, b, c) = SELECT (x, y, z)` etc. I know MySQL supports this for "in" lookups, but I don't know about "exact". I'll need to investigate, but I don't think handling non-direct rhs will be possible in all backends.

Reply:
Makes sense. If it's too much at first we should at least raise a clear exception that it's not supported instead of a crash or silently executing the queryset.